### PR TITLE
[17.0][MIG] account_commission: Migration to 17.0

### PR DESCRIPTION
**Pull request by ambientWave**
I upgraded the 'account_commission' module to be installed on Odoo 17.0. More focus was given to upgrading xml files.

The demo files weren't upgraded and hence, an error during installation is warning user that demo data wasn't populated. However, the module can be used without demo data.

Python code needs to be tested thoroughly.

All contributions and comments are welcomed.

### CODE REVIEW

**Comment by ambientWave** [comment]
The following error is seen in Odoo and OCB tests:
```
odoo-addon-account_commission @ file:///__w/commission/commission/account_commission
+ pip install -r test-requirements.txt -c test-constraints.txt
WARNING: The directory '/github/home/.cache/pip' or its parent directory is not owned or is not writable by the current user. The cache has been disabled. Check the permissions and owner of that directory. If executing pip with sudo, you should use sudo's -H flag.
Looking in indexes: https://wheelhouse.odoo-community.org/oca-simple-and-pypi
ERROR: Could not find a version that satisfies the requirement odoo-addon-commission<17.1dev,>=17.0dev (from versions: 15.0.2.0.0, 15.0.2.0.0.2, 15.0.2.0.0.3, 15.0.2.0.0.4, 15.0.2.0.0.6, 15.0.2.0.0.9, 15.0.2.1.0, 15.0.2.1.0.2, 15.0.2.2.0, 15.0.2.2.0.1, 15.0.2.2.0.2, 15.0.2.2.0.3, 15.0.2.2.0.5, 15.0.3.0.0, 15.0.3.1.0, 16.0.1.0.0.3, 16.0.1.0.0.4, 16.0.1.0.0.6, 16.0.1.0.0.8, 16.0.1.1.0, 16.0.1.1.1, 16.0.1.2.0, 16.0.1.2.0.1, 16.0.1.2.0.2, 16.0.1.2.0.4, 16.0.1.2.0.6, 16.0.1.2.0.7, 16.0.1.2.0.10, 16.0.1.2.0.12, 16.0.1.2.0.13, 16.0.1.2.0.15, 16.0.1.2.0.16, 16.0.1.2.0.19, 16.0.2.0.0, 16.0.2.1.0)
ERROR: No matching distribution found for odoo-addon-commission<17.1dev,>=17.0dev
Error: Process completed with exit code 1.
```
@sbidoul Any guess how to tackle this?

**Comment by ambientWave** [comment]
@pedrobaeza Currently, there's a method named 'action_switch_invoice_into_refund_credit_note' that is called from 'account.move' model which seems to exist no more.

As a quick solution, I had ported the method from Odoo 16.0 and pasted it inside the inherited account.move. Odoo tests are passed.

But I have noticed that this method has a new name and has been slightly changed in Odoo 17.0:
```
    def action_switch_move_type(self):
        if any(move.posted_before for move in self):
            raise ValidationError(_("You cannot switch the type of a posted document."))
        if any(move.move_type == "entry" for move in self):
            raise ValidationError(_("This action isn't available for this document."))

        for move in self:
            in_out, old_move_type = move.move_type.split('_')
            new_move_type = f"{in_out}_{'invoice' if old_move_type == 'refund' else 'refund'}"
            move.name = False
            move.write({
                'move_type': new_move_type,
                'partner_bank_id': False,
                'currency_id': move.currency_id.id,
            })
            if move.amount_total < 0:
                move.write({
                    'line_ids': [
                        Command.update(line.id, {'quantity': -line.quantity})
                        for line in move.line_ids
                        if line.display_type == 'product'
                    ]
                })
```

instead of this in Odoo 16.0:
```

    def action_switch_invoice_into_refund_credit_note(self):
        if any(move.move_type not in ('in_invoice', 'out_invoice') for move in self):
            raise ValidationError(_("This action isn't available for this document."))

        for move in self:
            move.write({
                'move_type': move.move_type.replace('invoice', 'refund'),
                'partner_bank_id': False,
                'currency_id': move.currency_id.id,
            })
            if move.amount_total < 0:
                move.write({
                    'line_ids': [
                        Command.update(line.id, {'quantity': -line.quantity})
                        for line in move.line_ids
                        if line.display_type == 'product'
                    ]
                })
```

What would you recommend?

**Comment by pedrobaeza** [comment]
You have to look for the equivalence of that method in this new version. That method was called for converting a normal invoice into a refund one and vice versa. How is this done now in v17? Once you find it, you have to make it work the same.

**Comment by ambientWave** [comment]
```
    def action_switch_move_type(self):
        if any(move.posted_before for move in self):
            raise ValidationError(_("You cannot switch the type of a posted document."))
        if any(move.move_type == "entry" for move in self):
            raise ValidationError(_("This action isn't available for this document."))

        for move in self:
            in_out, old_move_type = move.move_type.split('_')
            new_move_type = f"{in_out}_{'invoice' if old_move_type == 'refund' else 'refund'}"
            move.name = False
            move.write({
                'move_type': new_move_type,
                'partner_bank_id': False,
                'currency_id': move.currency_id.id,
            })
            if move.amount_total < 0:
                move.write({
                    'line_ids': [
                        Command.update(line.id, {'quantity': -line.quantity})
                        for line in move.line_ids
                        if line.display_type == 'product'
                    ]
                })
```

@pedrobaeza This is the method that we're looking for. I would paste the contents thereof under the old inherited method name. Is this approach any good?

**Comment by pedrobaeza** [comment]
You mustn't copy/paste anything. Just inherit and adapt the proper method.

**Comment by ambientWave** [comment]
@pedrobaeza Adapted successfully!

**Comment by Tisho99** [comment]
Hi @ambientWave 

I have also been reviewing this PR locally, and looks nice.

However, can you do the following to complete the migration?:
- Review and/or apply the changes i have made [on my branch](https://github.com/sygel-technology/commission/commit/8214b50a3c3a74024f9447627be19c892f35aeff). Some are upgrades and other [will be required for the migration](https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-17.0)
- Squash all the commits and execute pre-commit to pass the first check

With this changes i think the PR will be ready

Thank you!

**Comment by ambientWave** [comment]
Hey @Tisho99

I have applied the requested changes. Could you check out the PR and review those?

**Comment by Tisho99** [comment]
@ambientWave 

Yes, thank you.

Let me some time, i think tomorrow i will have it reviewed

File: account_commission/views/account_move_views.xml [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+                    />
+                </button>
+            </xpath>
+            <xpath
+                expr="//field[@name='invoice_line_ids']/tree//field[@name='price_subtotal']"
+                position="after"
+            >
+                <field name="settlement_id" invisible="1" />
+                <field name="commission_free" invisible="1" />
+                <field name="any_settled" invisible="1" />
```
Comment (Tisho99) [inline account_commission/views/account_move_views.xml]: ``
```suggestion
                <field name="settlement_id" column_invisible="1" />
                <field name="commission_free" column_invisible="1" />
                <field name="any_settled" column_invisible="1" />
```

**Review by Tisho99 — CHANGES_REQUESTED** [review]

File: account_commission/views/account_move_views.xml (hunk 1 above)
Comment (ambientWave) [reply account_commission/views/account_move_views.xml]: @Tisho99

Applied the changes!

**Review by Tisho99 — APPROVED** [review]
LGTM! Functional and technical review

**Comment by Tisho99** [comment]
Hi @ambientWave 

Your migration of commission was merged 

Can you remove the test-requirements.txt commit?

Thank you!

**Comment by ambientWave** [comment]
@Tisho99

test-requirements.txt commit was dropped. All checks are passed

**Review by Jaimermaccione — APPROVED** [review]
LGTM! Functional review on runboot.

I'd like to comment on two minor issues :

- The README should correctly locate the "Agent" checkbox under the "Sales and Purchases" tab; it is currently found below the "Tax ID."

- Clarification is needed on the "Regenerate Agents" button. From my experience, it must be clicked to apply new commission type configuration and settle commissions after changes are made. Perhaps updating the README and providing clear instructions on using the "Regenerate Agents" button, based on user experiences, will enhance the module's usability.

File: account_commission/models/account_move.py [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+                _("You can't cancel an invoice with settled lines"),
+            )
+        self.mapped("line_ids.settlement_id").write({"state": "except_invoice"})
+        return super().button_cancel()
+
+    def recompute_lines_agents(self):
+        self.mapped("invoice_line_ids").recompute_agents()
+
+    @api.model
+    def fields_view_get(
```
Comment (pedrobaeza) [inline account_commission/models/account_move.py]: Replace this by `_get_view`

Comment (ambientWave) [reply account_commission/models/account_move.py]: @pedrobaeza Method name was modified!

File: account_commission/models/account_move.py [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+
+    def recompute_lines_agents(self):
+        self.mapped("invoice_line_ids").recompute_agents()
+
+    @api.model
+    def _get_view(self, view_id=None, view_type="form", toolbar=False, submenu=False):
+        """Inject in this method the needed context for not removing other
+        possible context values.
+        """
+        res = super().fields_view_get(
```
Comment (Tisho99) [inline account_commission/models/account_move.py]:
```suggestion
        res = super()._get_view(
```

Comment (Tisho99) [reply account_commission/models/account_move.py]: Have you forgotten this?

**Review by Tisho99 — CHANGES_REQUESTED** [review]

**Review by Tisho99 — APPROVED** [review]

**Comment by Tisho99** [comment]
Sorry @ambientWave 

The get_view call is still no correct. Can you see [this example](https://github.com/odoo/odoo/commit/b03c227e885efa4ccdcad43ccb56ee10371b9284#diff-9d31094bdf7f812b4cc4adc3498384e42cc6477ada91f39ff7c33decf8b7268fR23)?

**Review by Tisho99 — CHANGES_REQUESTED** [review]
Sorry @ambientWave

Im afraid the get_view call is still no correct. Can you see [this example](https://github.com/odoo/odoo/commit/b03c227e885efa4ccdcad43ccb56ee10371b9284#diff-9d31094bdf7f812b4cc4adc3498384e42cc6477ada91f39ff7c33decf8b7268fR23)?

It is in the [v16 migration guide](https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-16.0)

**Comment by Tisho99** [comment]
Hi @ambientWave 

I have done a supersede. If you have time you can git cherry-pick the last commit of my branch and i will close the supersede.

https://github.com/sygel-technology/commission/commits/17.0-mig-account_commission_backup/

**Comment by jimhuds** [comment]
Thanks both for upgrading this. @Tisho99 I just tried your edits and I'm getting a server error:

`RPC_ERROR
Odoo Server Error
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/odoo/tools/cache.py", line 99, in lookup
    r = d[key]
  File "<decorator-gen-5>", line 2, in __getitem__
  File "/usr/lib/python3/dist-packages/odoo/tools/func.py", line 87, in locked
    return func(inst, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/odoo/tools/lru.py", line 34, in __getitem__
    a = self.d[obj]
KeyError: ('account.move', <function Model._get_view_cache at 0xf5ce83a80b80>, (585, 'tree', None, 'en_GB', None))`

This happens when I enable the account_commission module and try to view any invoices (list or form)

**Comment by pedrobaeza** [comment]
Superseded by #524

**Pull request closed without merging**